flocker/node/agents/ebs.py

```diff
@@ -152,6 +201,62 @@ def compute_instance_id(self):
         """
         return get_instance_metadata()['instance-id'].decode("ascii")
 
+    def _get(self, blockdevice_id):
+        """
+        Lookup BlockDeviceVolume for given blockdevice_id.
+
+        :param unicode blockdevice_id: ID of a blockdevice that needs lookup.
+
+        :returns BlockDeviceVolume for the given input id.
+        """
+        for volume in self.list_volumes():
+            if volume.blockdevice_id == blockdevice_id:
+                return volume
+        raise UnknownVolume(blockdevice_id)
+
+    def _get_ebs_volume(self, blockdevice_id):
+        """
+        Lookup EBS Volume information for a given blockdevice_id.
+
+        :param unicode blockdevice_id: ID of a blockdevice that needs lookup.
+
+        :returns boto.ec2.volume.Volume for the input id. ``None`` if
+            no boto.ec2.volume.Volume was found for the given id.
+        """
+        for volume in self.connection.get_all_volumes():
+            if volume.id == blockdevice_id:
+                # Sync volume for uptodate metadata
+                volume.update()
+                return volume
+        return None
+
+    def _next_device(self, instance_id):
+        """
+        Get the next available EBS device name for a given EC2 instance.
+        Algorithm:
+        1. Get all ``Block devices`` currently in use by given instance:
+            a) List all volumes visible to this instance.
+            b) Gather device IDs of all devices attached to (a).
+        2. Devices available for EBS volume usage are ``/dev/sd[f-p]``.
+           Find the first device from this set that is currently not
+           in use.
+
+        :param unicode instance_id: EC2 instance ID.
+
+        :returns unicode file_name: available device name for attaching
+            EBS volume.
+        :returns ``None`` if suitable EBS device names on this EC2
+            instance are currently occupied.
+        """
+        volumes = self.connection.get_all_volumes()
+        devices = [v.attach_data.device for v in volumes
+                   if v.attach_data.instance_id == instance_id]
+        for prefix in ['f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p']:
```

Comment:
Strings are iterable.  I'm not sure if this would be better or worse as `for prefix in "fghijklmnopq":`.  It'd be shorter, anyway. :wink:   One reason to change might be that these should probably all be explicit unicode or byte string literals and `b"fghijklmnopq"` is a lot easier than `b"f", b"g", ...`.

Speaking of, we generally treat filesystem paths as bytes (because we don't know the filesystem encoding so treating them as unicode is harder).  There are some exceptions and often it doesn't matter (as in this case where the paths are always pure ASCII which is easy to convert between unicode and bytes).  `FilePath` doesn't support `unicode`, though, so that's a good reason to stick with bytes.